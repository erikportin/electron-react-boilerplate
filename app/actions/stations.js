// @flow
export const ADD_STATION = 'ADD_STATION';
export const REMOVE_STATION = 'REMOVE_STATION';

type actionType = {
  +type: string
};

export function add() {
  return {
    type: ADD_STATION
  };
}

export function remove() {
  return {
    type: REMOVE_STATION
  };
}
