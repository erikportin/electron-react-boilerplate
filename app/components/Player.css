.container {
  height: 100vh;
}

.title {
  float: right;
  margin: 0;

}

.btn{
  float: left;
  padding: 0.5rem;
  display: block;
}

.header{
  padding: 1rem;
  overflow: hidden;
}
