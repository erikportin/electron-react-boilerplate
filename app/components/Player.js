// @flow
import React, { Component } from 'react';
import { ipcRenderer } from 'electron';
import { Link } from 'react-router-dom';
import styles from './Player.css';
import type { stationStateType } from '../reducers/stations';

type Props = {
  activate: () => void,
  deactivate: () => void,
  playPauseKeyActive: boolean,
  station: stationStateType
};

export default class Player extends Component<Props> {
  props: Props;

  constructor(props) {
    super(props);
    this.injectScript = this.injectScript.bind(this);
  }

  componentDidMount() {
    const {
      activate
    } = this.props;

    ipcRenderer.on('mediaplaypause', () => {
      activate();
    });
  }

  shouldComponentUpdate() {
    return true;
  }

  componentDidUpdate() {
    if (this.props.playPauseKeyActive) {
      this.injectScript();
      this.props.deactivate();
    }
  }

  injectScript() {
    const { script } = this.props.station;

    if (this.webview.executeJavaScript) {
      this.webview.executeJavaScript(script);
    }
  }

  render() {
    const { url, name } = this.props.station;

    return (
      <div>
        <h1>{name}</h1>
        <Link to="/">Back</Link>

        <webview
          data-tid="player"
          src={url}
          disablewebsecurity="true"
          className={styles.container}
          ref={(webview) => { this.webview = webview; }}
        />
      </div>
    );
  }
}
