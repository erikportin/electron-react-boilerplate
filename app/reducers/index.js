// @flow
import { combineReducers } from 'redux';
import { routerReducer as router } from 'react-router-redux';
import playPauseKeyActive from './mediaKeys';

const rootReducer = combineReducers({
  playPauseKeyActive,
  router,
});

export default rootReducer;
