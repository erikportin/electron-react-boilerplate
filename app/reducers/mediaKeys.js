// @flow
import { PLAY_PAUSE_DEACTIVATED, PLAY_PAUSE_ACTIVATED } from '../actions/mediaKeys';

export type playPauseKeyActiveStateType = {
  +playPauseKeyActive: boolean
};

type actionType = {
  +type: string
};

export default function playPauseKeyActive(state: boolean = false, action: actionType) {
  switch (action.type) {
    case PLAY_PAUSE_DEACTIVATED:
      return false;
    case PLAY_PAUSE_ACTIVATED:
      return true;
    default:
      return state;
  }
}
