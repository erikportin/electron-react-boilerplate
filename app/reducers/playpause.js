// @flow
import { PLAY_PAUSE_DEACTIVATED, PLAY_PAUSE_ACTIVATED } from '../actions/playpause';

export type playpauseStateType = {
  +counter: number
};

type actionType = {
  +type: string
};

export default function playpause(state: boolean = false, action: actionType) {
  switch (action.type) {
    case PLAY_PAUSE_DEACTIVATED:
      return false;
    case PLAY_PAUSE_ACTIVATED:
      return true;
    default:
      return false;
  }
}
