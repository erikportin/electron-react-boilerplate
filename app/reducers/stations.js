import { persistentReducer } from 'redux-pouchdb';
import { REMOVE_STATION, ADD_STATION } from '../../app/actions/stations';

export type stationStateType = {
  name: url,
  url: string,
  script: string
};

export type stationsStateType = {
  +stations: Array<stationStateType>
};

type actionType = {
  +type: string
};

export const initialState = [
  {
    url: 'http://www.bbc.co.uk/programmes/b09s37vt',
    name: 'BBC 6',
    script: `var iFrame = document.querySelectorAll('.episode-playout iframe')[0].contentDocument;
    var pauseBtn = iFrame.querySelectorAll('.p_pauseButton')[0];
    var playBtn = iFrame.querySelectorAll('.p_playButton')[0];
    var startBtn = iFrame.querySelectorAll('.p_button')[0];
    
    
    if(pauseBtn){
        pauseBtn.click()
    }
    else if(playBtn){
        playBtn.click()
    }
    else if(startBtn){
        startBtn.click()
    }`
  }
];

const stations = (state = initialState, action) => {
  switch (action.type) {
    case ADD_STATION:
      return [
        ...state,
        action.station
      ];
    case REMOVE_STATION:
      return [
        ...state.slice(0, action.id),
        ...state.slice(action.id + 1)
      ];
    default:
      return state;
  }
};

export default persistentReducer(stations);
