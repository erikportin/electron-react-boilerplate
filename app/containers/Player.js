import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import Player from '../components/Player';
import * as MediaKeysAction from '../actions/mediaKeys';

function mapStateToProps(state) {
  return {
    playPauseKeyActive: state.playPauseKeyActive
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(MediaKeysAction, dispatch);
}

export default connect(mapStateToProps, mapDispatchToProps)(Player);
