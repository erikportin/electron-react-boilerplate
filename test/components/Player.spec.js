import { spy } from 'sinon';
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import Player from '../../app/components/Player';
import { initialState as stations } from '../../app/reducers/stations';

Enzyme.configure({ adapter: new Adapter() });


function setup() {
  const actions = {
    deactivate: spy(),
    activate: spy()
  };

  const component = mount(<Player playPauseKeyActive={false} station={stations[0]} {...actions} />);

  return {
    component,
    actions,
    webview: component.find('[data-tid="webview"]')
  };
}

describe('Player component', () => {
  it('should display webview with src', () => {
    const { webview } = setup();
    expect(webview.prop('src')).toEqual('http://www.bbc.co.uk/programmes/b09s37vt');
  });

  it('should handle play pause action', () => {
    const { actions, component } = setup();
    component.setProps({ playPauseKeyActive: true });
    expect(actions.deactivate.called).toBe(true);
  });

  it('should match exact snapshot', () => {
    const { component } = setup();

    const tree = renderer
      .create(component)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
