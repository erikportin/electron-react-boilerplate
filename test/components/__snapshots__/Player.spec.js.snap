// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player component should match exact snapshot 1`] = `
<div
  data-tid="player"
>
  <div
    className="header"
  >
    <div>
      <i
        className="fa fa-chevron-circle-left"
      />
    </div>
    <h1
      className="title"
    >
      BBC 6
    </h1>
  </div>
  <webview
    className="container"
    data-tid="webview"
    disablewebsecurity="true"
    src="http://www.bbc.co.uk/programmes/b09s37vt"
  />
</div>
`;
